include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,681 @@
+auto
+__parallel_histogram_select_kernel(_ExecutionPolicy&& __exec, const sycl::event& __init_e, _Range1&& __input,
+                                   _Range2&& __bins, _BinHashMgr __binhash_manager)
+{
+    using _private_histogram_type = ::std::uint16_t;
+    using _local_histogram_type = ::std::uint32_t;
+    using _extra_memory_type = typename _BinHashMgr::_extra_memory_type;
+
+    const auto __num_bins = __bins.size();
+    ::std::size_t __max_wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+    ::std::uint16_t __work_group_size = ::std::min(::std::size_t(1024), __max_wgroup_size);
+
+    auto __local_mem_size = __exec.queue().get_device().template get_info<sycl::info::device::local_mem_size>();
+    constexpr ::std::uint8_t __max_work_item_private_bins = 16 / sizeof(_private_histogram_type);
+
+    // if bins fit into registers, use register private accumulation
+    if (__num_bins <= __max_work_item_private_bins)
+    {
+        return __future(
+            __histogram_general_registers_local_reduction<__iters_per_work_item, __max_work_item_private_bins>(
+                ::std::forward<_ExecutionPolicy>(__exec), __init_e, __work_group_size, ::std::forward<_Range1>(__input),
+                ::std::forward<_Range2>(__bins), __binhash_manager));
+    }
+    // if bins fit into SLM, use local atomics
+    else if (__num_bins * sizeof(_local_histogram_type) +
+                 __binhash_manager.get_required_SLM_elements() * sizeof(_extra_memory_type) <
+             __local_mem_size)
+    {
+        return __future(__histogram_general_local_atomics<__iters_per_work_item>(
+            ::std::forward<_ExecutionPolicy>(__exec), __init_e, __work_group_size, ::std::forward<_Range1>(__input),
+            ::std::forward<_Range2>(__bins), __binhash_manager));
+    }
+    else // otherwise, use global atomics (private copies per workgroup)
+    {
+        //Use __iters_per_work_item here as a runtime parameter, because only one kernel is created for
+        // private_global_atomics with a variable number of iterations per workitem. __iters_per_work_item is just a
+        // suggestion which but global memory limitations may increase this value to be able to fit the workgroup
+        // private copies of the histogram bins in global memory.  No unrolling is taken advantage of here because it
+        // is a runtime argument.
+        return __future(__histogram_general_private_global_atomics(
+            ::std::forward<_ExecutionPolicy>(__exec), __init_e, __iters_per_work_item, __work_group_size,
+            ::std::forward<_Range1>(__input), ::std::forward<_Range2>(__bins), __binhash_manager));
+    }
+}
+
+template <typename _Name>
+struct __hist_fill_zeros_wrapper
+{
+};
+
+template <typename _ExecutionPolicy, typename _Iter1, typename _Size, typename _IdxHashFunc, typename _Iter2>
+void
+__parallel_histogram(_ExecutionPolicy&& __exec, _Iter1 __first, _Iter1 __last, _Size __num_bins, _IdxHashFunc __func,
```

Comment:
I think it would be best to add histogram to the experimental async and range APIs in a follow up PR.  

Unless you see this as a requirement for approval for this PR, I will make an issue to add this as a follow up PR.